ktesting/testinglogger.go

```diff
@@ -250,8 +266,12 @@ func (l tlogger) fallbackLogger() logr.Logger {
 	logger = logger.WithCallDepth(l.shared.callDepth + 1)
 
 	if !l.shared.goroutineWarningDone {
-		logger.WithCallDepth(1).Error(nil, "WARNING: test kept at least one goroutine running after test completion", "callstack", string(dbg.Stacks(false)))
-		l.shared.goroutineWarningDone = true
+		duration := time.Since(l.shared.stopTime)
+		if duration > stopGracePeriod {
```

Comment:
This seems will bury the real goroutine leakage if the duration is less than stopGracePeriod, do I understand right here? 

Another point I can't get here is if the test completes, why will it wait for another 10s?

Reply:
If the "leaked" goroutine stops after less than `stopGracePeriod`, then it wasn't really a leak. It was "just" a test that didn't block and wait for goroutines at the end. This is very common.

Reply:
Totally misunderstood the meaning here. I'm not quite understand how it works but I still have a silly question like: Will we wait and detect the goroutine leakage after test completes? Or why the duration from *stop* will be greater than `stopGracePeriod`.

Reply:
No, we don't wait. That is something that [uber/goleak](https://github.com/uber-go/goleak) does, but not this code. This code is merely reporting on leaked goroutines if and when it detects them, which is when they keep logging.

The `stopGracePeriod` is the grace period that goroutines have before their log output triggers this warning. I can rename it if you have a better suggestion.

Reply:
I recalled goleak has maxRetries, then can't we just configure the goleak? It's configurable now. Both the retry times and sleep duration. I think somehow the same as `stopGracePeriod`. Will this lead to other problems?

Reply:
This is klog. We cannot depend on goleak here, nor do I want to re-implement it.

The reason why this warning exists is because the per-test logger has to do something. It cannot call `testing.T.Log` after the test that `testing.T` belongs to has terminated, it would crash - I know, because that has happened initially.

When it detects that a test has stopped, it handles output differently and writes to stderr. Issuing this warning seemed usedful, but printing it immediately has confused Kubernetes developers because the goroutine hasn't really leaked. This PR is an attempt to keep the warning for goroutines that are more likely to be real leaks while not printing it for goroutines which terminate quickly after test completion.

Reply:
> keep the warning for goroutines that are more likely to be real leaks while not printing it for goroutines which terminate quickly after test completion.

Definitely agree.

> We cannot depend on goleak here

Yes for klog... I mean for callers, like test framework. I suggested in this way is just because the goroutine leakage detection duration is mostly decided by goleak, the maxRetries(default to 20), maxSleep(default to 100ms) specifically, the `duration ~= maxRetries * (maxSleep + filterStacks)`, have no idea whether 10s fits here.  However, the feature is reverted... https://github.com/uber-go/goleak/pull/116, so no way to configure now. 

As a user of klog, happy with either way to avoid confusion.

Reply:
> Definitely agree.

So LGTM?

> I mean for callers, like test framework.

Leak detection with goleak only works when tests run sequentially or in TestMain. I'm not sure how useful it is in an E2E test suite, because the actual code under test runs elsewhere. For integration and unit tests, individual test owners have to integrate goleak. Some apiserver-based tests have it.

> have no idea whether 10s fits here

So this is about how klog will interact with goleak. I don't have a strong opinion here. I picked 10 seconds because it should be long enough to no produce "false positives" for the "I am shutting down after test completion" goroutines and short enough to give feedback for others. Because it's not causing a test failure, it can be lower than the goleak timeouts.

Reply:
/lgtm

Reply:
are these 10s added to the test duration? 10s sounds like a lot to me

Reply:
No, they are not added. At the time that this gets called, the test has already completed.